app/plugin.go

```diff
@@ -394,3 +398,48 @@ func (a *App) GetPlugins() (*model.PluginsResponse, *model.AppError) {
 
 	return resp, nil
 }
+
+// notifyPluginEnabled notifies web clients for all cluster peers if
+// manifest version is same across all peers, including self.
+func (a *App) notifyPluginEnabled(manifest *model.Manifest) *model.AppError {
+	if manifest.HasClient() {
+		var statuses model.PluginStatuses
+		expectedPeers := 1
+		actualPeers := 0
+
+		if a.Cluster != nil {
+			var err *model.AppError
+			statuses, err = a.Cluster.GetPluginStatuses()
+			if err != nil {
+				return err
+			}
+			if infos := len(a.Cluster.GetClusterInfos()); infos > 0 {
+				expectedPeers = infos
+			}
+		}
+
+		localStatus, err := a.GetPluginStatus(manifest.Id)
+		if err != nil {
+			return err
+		}
+		statuses = append(statuses, localStatus)
+
+		for _, status := range statuses {
+			// Count consistent peers
+			isPeerConsistent := status.Version == manifest.Version &&
+				localStatus.State == model.PluginStateRunning && localStatus.State == status.State
+			if status.PluginId == manifest.Id && isPeerConsistent {
+				actualPeers++
+			}
+		}
+
+		// Notify all cluster peer clients.
+		if actualPeers == expectedPeers {
+			message := model.NewWebSocketEvent(model.WEBSOCKET_EVENT_PLUGIN_ENABLED, "", "", "", nil)
+			message.Add("manifest", manifest.ClientManifest())
+			a.Publish(message)
```

Comment:
This can happen multiple times because we are not making sure we are the last one to be in the correct state. We just make sure that everyone else is.  I don't think this is a problem though.

Reply:
True. I've updated webapp to make `plugin_enabled` and `plugin_disabled` events idempotent, to guard against multiple events going out.
https://github.com/mattermost/mattermost-webapp/pull/3425